cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go

```diff
@@ -23,6 +23,7 @@ package v1beta4
 
 import (
 	runtime "k8s.io/apimachinery/pkg/runtime"
+	v1 "k8s.io/kubernetes/pkg/apis/core/v1"
```

Comment:
As I tested, this is the key import that is related.

Reply:
@pacoxu as mentioned the kubernetes/pkg import is _not allowed_ in kubeadm!

@chendave we must find out how to solve it. maybe it's a generator bug.
if we cannot solve it we can try avoiding the defaulting `[]...{}`

if we are out of ideas, we must revert the PR.

Reply:
We must drop this `import` first.
In the past, we have made a lot of efforts to do this, and we cannot go back.
But I'm confused, why didn't the [.import-restrictions](https://raw.githubusercontent.com/kubernetes/kubernetes/master/cmd/kubeadm/.import-restrictions) prevent PRs from being merged?

Reply:
perhaps the import-boss tool is not running or buggy.
it can be tested with a simple PR for cmd/kubeadm or other packages that have an .import-restrictions file.

Reply:
I tested it locally, and it seems that the generated code(`zz_generated.defaults.go`) won't be checked...

Reply:
It makes sense. So generated code are skipped for that import restrictions.

Reply:
> 2\. maybe there is already `+foo` directive for it, like `+optional`.

```
~/go/src/k8s.io/kubernetes$ git diff
diff --git a/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go b/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go
index 314be0b30c7..38819c814f4 100644
--- a/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go
+++ b/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go
@@ -74,6 +74,7 @@ func SetDefaults_InitConfiguration(obj *InitConfiguration) {
 }
 
 // SetDefaults_ClusterConfiguration assigns default values for the ClusterConfiguration
+// +k8s:defaulter-gen=covers
 func SetDefaults_ClusterConfiguration(obj *ClusterConfiguration) {
        if obj.KubernetesVersion == "" {
                obj.KubernetesVersion = DefaultKubernetesVersion
```

results in no import of `k8s.io/kubernetes/pkg/api`

> // An existing defaulter method (`SetDefaults_TYPE`) can provide the
> // comment tag:
> //
> //	// +k8s:defaulter-gen=covers
> //
> // to indicate that the defaulter does not or should not call any nested
> // defaulters.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/defaulter-gen/main.go#L35-L41

Reply:
result diff:

```
diff --git a/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go b/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go
index b36b0f0f952..d2409b0de6a 100644
--- a/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go
+++ b/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go
@@ -23,7 +23,6 @@ package v1beta4
 
 import (
        runtime "k8s.io/apimachinery/pkg/runtime"
-       v1 "k8s.io/kubernetes/pkg/apis/core/v1"
 )
 
 // RegisterDefaults adds defaulters functions to the given scheme.
@@ -39,41 +38,6 @@ func RegisterDefaults(scheme *runtime.Scheme) error {
 
 func SetObjectDefaults_ClusterConfiguration(in *ClusterConfiguration) {
        SetDefaults_ClusterConfiguration(in)
-       if in.Etcd.Local != nil {
-               for i := range in.Etcd.Local.ExtraEnvs {
-                       a := &in.Etcd.Local.ExtraEnvs[i]
-                       if a.ValueFrom != nil {
-                               if a.ValueFrom.FieldRef != nil {
-                                       v1.SetDefaults_ObjectFieldSelector(a.ValueFrom.FieldRef)
-                               }
-                       }
-               }
-       }
-       SetDefaults_APIServer(&in.APIServer)
-       for i := range in.APIServer.ControlPlaneComponent.ExtraEnvs {
-               a := &in.APIServer.ControlPlaneComponent.ExtraEnvs[i]
-               if a.ValueFrom != nil {
-                       if a.ValueFrom.FieldRef != nil {
-                               v1.SetDefaults_ObjectFieldSelector(a.ValueFrom.FieldRef)
-                       }
-               }
-       }
-       for i := range in.ControllerManager.ExtraEnvs {
-               a := &in.ControllerManager.ExtraEnvs[i]
-               if a.ValueFrom != nil {
-                       if a.ValueFrom.FieldRef != nil {
-                               v1.SetDefaults_ObjectFieldSelector(a.ValueFrom.FieldRef)
-                       }
-               }
-       }

```


questions:
- would that be a problem for us in the future? probably no, because if we want a defaulter for some field in ClusterConfiguration we can explicitly call it in SetDefaults_ClusterConfiguration()
- can this missing defaulting for `a.ValueFrom.FieldRef` be a problem? i think it just sets that field selector to "v1", but technically the kube-apiserver must also do this defaulting when the mirror Pod (of the static pod) is created, right?

Reply:
okay, I like this approach, it's more neat to look!